### ISSUE TITLE
Support for desktop accessibility APIs

### ISSUE THREAD

**Issue by gspencergoog** [flutter/flutter#23601]
For some platforms (ChromeOS, Fuschia), in order to have useful accessibility information, we need to be connected to the platform's accessibility API (at least to know when we need to turn on the accessibility modes in Flutter).

**Comment by stuartmorgan** [comment]
This is definitely something that's of interest to flutter-desktop-embedding as well (e.g., macOS VoiceOver support)

**Comment by camlorn** [comment]
For what it's worth, Windows is almost certainly UI Automation.  The other two options are MSAA and IA2.  MSAA is both limited and deprecated (in the no longer maintained sense, though I might be wrong) and IA2 is an unofficial third-party extension.  IA2 might be feasible, if Microsoft adds support to Narrator, as afaik that's what Chrome uses and they're switching to Chromium, but it's not an official Microsoft technology.

In all 3 cases, you do it by responding to the `WM_GETOBJECT` window message and returning a COM interface via some marshalling magic I don't remember the details of offhand.  The difference between these APIs is which COM interface(s) you're implementing.

**Comment by stuartmorgan** [comment]
One potential option is for us to extract https://cs.chromium.org/search/?q=ui/accessibility/platform from Chromium and re-use it to cover all three desktop platforms. Then we'd only need to do the binding between that API and the embedder.h accessibility API, and then it would take care of binding to the OS-specific APIs.

It'll need to be untangled from base, but assuming it meets our needs otherwise (I haven't looked at how well it aligns to the Flutter abstractions), replacing the necessary parts of base should be substantially simpler than writing N binding layers.

**Comment by camlorn** [comment]
If there's an accessibility API binding that can be used across multiple projects, that would be valuable in itself, so if you go that route it might be worth considering putting it in a separate repo as its own library or somesuch assuming that's not difficult to do.  Lots of other GUI frameworks have gone the GPU rendering route and the answer for almost all of them is to write n binding layers at the moment.

**Comment by mivoligo** [comment]
Is the keyboard navigation inside Linux apps depending on this issue?

**Comment by gspencergoog** [comment]
@mivoligo No, regular keyboard navigation (tab traversal, etc.) is separate from the accessibility navigation support, and already works.

**Comment by mivoligo** [comment]
@gspencergoog Thanks for your answer. It does not work for me on Ubuntu. Could you check the bug #63078 please?

**Comment by francipvb** [comment]
Hi,

I'minterested on this issue. How it goes?

**Comment by cbracken** [comment]
Verified with Linux GTK embedding this morning.
